cabal-install/Distribution/Client/CmdSdist.hs

```diff
@@ -0,0 +1,289 @@
+    , readTargetSelectors, reportTargetSelectorProblems )
+import Distribution.Client.RebuildMonad
+    ( runRebuild )
+import Distribution.Client.Setup
+    ( ArchiveFormat(..), GlobalFlags(..) )
+import Distribution.Solver.Types.SourcePackage
+    ( SourcePackage(..) )
+import Distribution.Client.Types
+    ( PackageSpecifier(..), PackageLocation(..), UnresolvedSourcePackage )
+import Distribution.Client.DistDirLayout
+    ( DistDirLayout(..), defaultDistDirLayout )
+import Distribution.Client.ProjectConfig
+    ( findProjectRoot, readProjectConfig )
+
+import Distribution.Package
+    ( Package(packageId) )
+import Distribution.PackageDescription.Configuration
+    ( flattenPackageDescription )
+import Distribution.Pretty
+    ( prettyShow )
+import Distribution.ReadE
+    ( succeedReadE )
+import Distribution.Simple.Command
+    ( CommandUI(..), option, choiceOpt, reqArg )
+import Distribution.Simple.PreProcess
+    ( knownSuffixHandlers )
+import Distribution.Simple.Setup
+    ( Flag(..), toFlag, fromFlagOrDefault, flagToList, flagToMaybe
+    , optionVerbosity, optionDistPref, trueArg
+    )
+import Distribution.Simple.SrcDist
+    ( listPackageSources )
+import Distribution.Simple.Utils
+    ( die', notice )
+import Distribution.Types.ComponentName
+    ( ComponentName, showComponentName )
+import Distribution.Types.PackageName
+    ( PackageName, unPackageName )
+import Distribution.Verbosity
+    ( Verbosity, normal )
+
+import qualified Codec.Archive.Tar       as Tar
+import qualified Codec.Archive.Tar.Entry as Tar
+import qualified Codec.Archive.Zip       as Zip
+import qualified Codec.Compression.GZip  as GZip
+import Control.Exception
+    ( throwIO )
+import Control.Monad
+    ( when, forM, forM_ )
+import Control.Monad.Trans
+    ( liftIO )
+import Control.Monad.State.Lazy
+    ( StateT, modify, gets, evalStateT )
+import Control.Monad.Writer.Lazy
+    ( WriterT, tell, execWriterT )
+import Data.Bits
+    ( shiftL )
+import qualified Data.ByteString.Lazy.Char8 as BSL
+import Data.Either
+    ( partitionEithers )
+import Data.List
+    ( find, sortBy, nub )
+import System.Directory
+    ( getCurrentDirectory, setCurrentDirectory
+    , createDirectoryIfMissing )
+import System.FilePath
+    ( (</>), (<.>), normalise, takeDirectory )
+
+sdistCommand :: CommandUI SdistFlags
+sdistCommand = CommandUI
+    { commandName = "new-sdist"
+    , commandSynopsis = "Generate a source distribution file (.tar.gz)."
+    , commandUsage = \pname ->
+        "Usage: " ++ pname ++ " new-sdist [FLAGS]\n"
+    , commandDescription  = Just $ \_ ->
+        "Generates a tarball of a package suitable for upload to Hackage."
+    , commandNotes = Nothing
+    , commandDefaultFlags = defaultSdistFlags
+    , commandOptions = \showOrParseArgs ->
+        [ optionVerbosity
+            sdistVerbosity (\v flags -> flags { sdistVerbosity = v })
+        , optionDistPref
+            sdistDistDir (\dd flags -> flags { sdistDistDir = dd })
```

Comment:
This is claiming its default is `dist`, but it actually plops the tarball in `dist-newstyle`.

Reply:
That's #5380 and it's its own issue. `optionDistPref` needs fixing, but that's neither here nor there for this patchset.